### PR TITLE
[downloader/http] Make retrying more lenient

### PR DESCRIPTION
**Pull request by coletdjnz**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [X] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Makes the http downloader much more lenient on what it retries on, to survive any sort of brief network interruption etc. (given enough retries for such). 

We are thinking it is better to retry on more even though some may be unrecoverable, vs retrying on few, known and likely to be recoverable errors but missing many other cases.

Resolves https://github.com/yt-dlp/yt-dlp/issues/3056, https://github.com/yt-dlp/yt-dlp/issues/2071, https://github.com/yt-dlp/yt-dlp/issues/3034, https://github.com/yt-dlp/yt-dlp/issues/2969 and prob many upstream

Might also help sorting exceptions in #2861 simpler.

### CODE REVIEW

File: yt_dlp/downloader/http.py [hunk 1]
```diff
@@ -202,14 +197,16 @@ def establish_connection():
...
                     raise
                 raise RetryDownload(err)
-            except socket.timeout as err:
-                raise RetryDownload(err)
-            except socket.error as err:
-                if err.errno in (errno.ECONNRESET, errno.ETIMEDOUT):
-                    # Connection reset is no problem, just retry
+            except compat_urllib_error.URLError as err:
+                # TODO: What errors should we exclude?
+                if not any(isinstance(err.reason, et) for et in (ssl.CertificateError, )):
```
Comment (Jules-A) [inline yt_dlp/downloader/http.py:202]: Huh, shouldn't there be something after that comma?

Comment (coletdjnz) [reply yt_dlp/downloader/http.py:202]: I was going to add more but couldn't think of any. Just left it like that for now 😅

Comment (pukkandan) [reply yt_dlp/downloader/http.py:202]: isinstance takes multiple items
```py
                if not isinstance(err.reason, (ssl.CertificateError, ...)):
```

**Comment by coletdjnz** [comment]
- I notice that exceptions raised by reading the response may also occur during request, due to part of the response being read for headers etc. These are not wrapped by URLError.

- Not sure if we should exclude ssl.SSLCertificateError
- Any other exceptions we should exempt?
   - Keeping in mind for the network overhaul (#2861), having general set of errors instead of specifics makes it a lot easier to adapt exceptions from higher level backends :laughing:)

**Pull request merged**